### ISSUE TITLE
internetarchive's 'yarn add https://github.com/internetarchive/dweb-mirror' fails on "32-bit" Raspberry Pi OS 12 Bookworm

### ISSUE THREAD

**Issue by holta** [iiab/iiab#3641]
@mitra42 do you happen to know the cause here?

If the subject line is a full enough summary of the issue, should we announce that we no longer support internetarchive IIAB app on "32-bit" Raspberry Pi OS — e.g. if it's deemed no longer worth the trouble — as the "Bookworm" (i.e. Debian 12) version of Raspberry Pi OS goes mainline very soon in coming weeks?

Ironically despite being labelled as a "32-bit" OS, this Raspberry Pi OS that we've been testing for more than 5 months... _is running a 64-bit aarch64 kernel, whenever it detects a modern 64-bit CPU (OS's official new behavior, since mid-March 2023, raspberrypi/firmware#1795) !_

In any case here's the error, that is confirmed repeatable: ( e.g. after running `cd /opt/iiab/iiab` then `./runrole internetarchive` )

```
TASK [internetarchive : Run 'yarn add https://github.com/internetarchive/dweb-mirror' to download/populate /opt/iiab/internetarchive/node_modules (CAN TAKE ~5 MINUTES)] ********************************************************************
fatal: [127.0.0.1]: FAILED! => {"changed": true, "cmd": "yarn config set child-concurrency 1 && yarn add https://github.com/internetarchive/dweb-mirror", "delta": "0:00:31.894871", "end": "2023-09-11 15:42:16.379388", "msg": "non-zero return code", "rc": 127, "start": "2023-09-11 15:41:44.484517", "stderr": "warning https://github.com/internetarchive/dweb-mirror > multihashes > multibase@4.0.6: This module has been superseded by the multiformats module\nwarning https://github.com/internetarchive/dweb-mirror > @internetarchive/dweb-archive-dist > react > fbjs > core-js@1.2.7: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.\nwarning https://github.com/internetarchive/dweb-mirror > @futurepress/epubjs-reader > epubjs > @types/localforage@0.0.34: This is a stub types definition for localforage (https://github.com/localForage/localForage). localforage provides its own type definitions, so you don't need @types/localforage installed!\nerror /opt/iiab/internetarchive/node_modules/sharp: Command failed.\nExit code: 127\nCommand: (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)\nArguments: \nDirectory: /opt/iiab/internetarchive/node_modules/sharp\nOutput:\nsharp: Using cached /root/.npm/_libvips/libvips-8.13.3-linux-armv7.tar.br\nsharp: Integrity check passed for linux-armv7\nprebuild-install warn install No prebuilt binaries found (target=7 runtime=napi arch=arm libc= platform=linux)\n/bin/sh: 1: node-gyp: not found", "stderr_lines": ["warning https://github.com/internetarchive/dweb-mirror > multihashes > multibase@4.0.6: This module has been superseded by the multiformats module", "warning https://github.com/internetarchive/dweb-mirror > @internetarchive/dweb-archive-dist > react > fbjs > core-js@1.2.7: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.", "warning https://github.com/internetarchive/dweb-mirror > @futurepress/epubjs-reader > epubjs > @types/localforage@0.0.34: This is a stub types definition for localforage (https://github.com/localForage/localForage). localforage provides its own type definitions, so you don't need @types/localforage installed!", "error /opt/iiab/internetarchive/node_modules/sharp: Command failed.", "Exit code: 127", "Command: (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)", "Arguments: ", "Directory: /opt/iiab/internetarchive/node_modules/sharp", "Output:", "sharp: Using cached /root/.npm/_libvips/libvips-8.13.3-linux-armv7.tar.br", "sharp: Integrity check passed for linux-armv7", "prebuild-install warn install No prebuilt binaries found (target=7 runtime=napi arch=arm libc= platform=linux)", "/bin/sh: 1: node-gyp: not found"], "stdout": "yarn config v1.22.19\nsuccess Set \"child-concurrency\" to \"1\".\nDone in 0.13s.\nyarn add v1.22.19\ninfo No lockfile found.\n[1/4] Resolving packages...\n[2/4] Fetching packages...\n[3/4] Linking dependencies...\n[4/4] Building fresh packages...\ninfo Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.", "stdout_lines": ["yarn config v1.22.19", "success Set \"child-concurrency\" to \"1\".", "Done in 0.13s.", "yarn add v1.22.19", "info No lockfile found.", "[1/4] Resolving packages...", "[2/4] Fetching packages...", "[3/4] Linking dependencies...", "[4/4] Building fresh packages...", "info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command."]}
```

Same error is pasted in below in another form — some people find more readable:

> TASK [internetarchive : Run 'yarn add https://github.com/internetarchive/dweb-mirror' to download/populate /opt/iiab/internetarchive/node_modules (CAN TAKE ~5 MINUTES)] ********************************************************************
fatal: [127.0.0.1]: FAILED! => {"changed": true, "cmd": "yarn config set child-concurrency 1 && yarn add https://github.com/internetarchive/dweb-mirror", "delta": "0:00:31.894871", "end": "2023-09-11 15:42:16.379388", "msg": "non-zero return code", "rc": 127, "start": "2023-09-11 15:41:44.484517", "stderr": "warning https://github.com/internetarchive/dweb-mirror > multihashes > multibase@4.0.6: This module has been superseded by the multiformats module\nwarning https://github.com/internetarchive/dweb-mirror > @internetarchive/dweb-archive-dist > react > fbjs > core-js@1.2.7: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.\nwarning https://github.com/internetarchive/dweb-mirror > @futurepress/epubjs-reader > epubjs > @types/localforage@0.0.34: This is a stub types definition for localforage (https://github.com/localForage/localForage). localforage provides its own type definitions, so you don't need @types/localforage installed!\n**error /opt/iiab/internetarchive/node_modules/sharp: Command failed.\nExit code: 127**\nCommand: (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)\nArguments: \nDirectory: /opt/iiab/internetarchive/node_modules/sharp\nOutput:\nsharp: Using cached /root/.npm/_libvips/libvips-8.13.3-linux-armv7.tar.br\nsharp: Integrity check passed for linux-armv7\nprebuild-install warn install No prebuilt binaries found (target=7 runtime=napi arch=arm libc= platform=linux)\n/bin/sh: 1: node-gyp: not found", "stderr_lines": ["warning https://github.com/internetarchive/dweb-mirror > multihashes > multibase@4.0.6: This module has been superseded by the multiformats module", "warning https://github.com/internetarchive/dweb-mirror > @internetarchive/dweb-archive-dist > react > fbjs > core-js@1.2.7: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.", "warning https://github.com/internetarchive/dweb-mirror > @futurepress/epubjs-reader > epubjs > @types/localforage@0.0.34: This is a stub types definition for localforage (https://github.com/localForage/localForage). localforage provides its own type definitions, so you don't need @types/localforage installed!", "**error /opt/iiab/internetarchive/node_modules/sharp: Command failed.", "Exit code: 127"**, "Command: (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)", "Arguments: ", "Directory: /opt/iiab/internetarchive/node_modules/sharp", "Output:", "sharp: Using cached /root/.npm/_libvips/libvips-8.13.3-linux-armv7.tar.br", "sharp: Integrity check passed for linux-armv7", "prebuild-install warn install No prebuilt binaries found (target=7 runtime=napi arch=arm libc= platform=linux)", "/bin/sh: 1: node-gyp: not found"], "stdout": "yarn config v1.22.19\nsuccess Set \"child-concurrency\" to \"1\".\nDone in 0.13s.\nyarn add v1.22.19\ninfo No lockfile found.\n[1/4] Resolving packages...\n[2/4] Fetching packages...\n[3/4] Linking dependencies...\n[4/4] Building fresh packages...\ninfo Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.", "stdout_lines": ["yarn config v1.22.19", "success Set \"child-concurrency\" to \"1\".", "Done in 0.13s.", "yarn add v1.22.19", "info No lockfile found.", "[1/4] Resolving packages...", "[2/4] Fetching packages...", "[3/4] Linking dependencies...", "[4/4] Building fresh packages...", "info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command."]}

Full context of the Raspberry Pi 4 machine in question:

http://sprunge.us/5HRXSO?en

```
root@box:~# iiab-summary
install_date = 2023-09-11T06:16:28Z   Current TZ: EDT

iiab: 821f1448b, 50 PR's / 225 commits since tag 8.1-preview-2
    "CLARIF: apt-not-pip on 32-bit i386 for ansible-core"
    https://github.com/iiab/iiab   branch: master

iiab-admin-console: f085beb, 19 PR's / 62 commits since tag v0.6.2
    "Merge pull request #545 from tim-moody/0.6.6-job-outp"
    https://github.com/iiab/iiab-admin-console   branch: master

Raspberry Pi reference 2023-05-03
Generated using pi-gen, https://github.com/RPi-Distro/pi-gen, 47eee1f0ddcf8811559d51eea1c1bb48335e3e88, stage2
/etc/debian-version: 12.1
display-manager? inactive   Arch1: armhf   Arch2:
box 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64
Cortex-A72   CPU(s): 4   Mem: 1849 Swap: 1023
Raspberry Pi 4 Model B Rev 1.1
Temperature(s): 54530

Filesystem      Size  Used Avail Use% Mounted on
/dev/root        59G  8.5G   49G  15% /      ZIMs: 1 OER2Go: 0 Apps2B: 4

lo eth0 wlan0
openvpn_enabled: False
openvpn_handle: 183-raspios-armhf-aarch64-LARGE
192.168.0.183

root@box:~# node -v
v20.6.1
root@box:~# npm -v
9.8.1
```

Tangentially Related "32-bit" problems:

- #3479
- #3639
- #3640

**Comment by mitra42** [comment]
Hi @holta , The unfortunate thing with Javascript dependency versioning is that there is always BitRot ... its hard to interpret the warning above (as is usual with JS installs) to figure out what is actually happening, but it looks to me like node-gyp is not present, (I think that ends up as an executable) the dependency chain is HashStore->...level->...->node-gyp 

I just tried installing dweb-mirror on my Mac, and it failed for the same reason. So I think its bitrot in Level, let me try and shift to a newer version of level and see if that fixes it.

**Comment by mitra42** [comment]
Ok - I think I nailed it, meaning I could duplicate the problem (in level) on my mac, and no longer have that problem. 

I'm not sure how to let someone with a 32 bit version know to test it, but I dont' actually think this was architecture related since I could easily duplicate on my Mac. 

(I'm currently in Indonesia, limited time, bandwidth etc, and certainly no other machine to test it on)

**Comment by holta** [comment]
Thanks!

Question: Should I re-run this in a different way?

I just now ran this:

```
cd /opt/iiab/iiab
./runrole internetarchive
```

The result appears similar-if-not-the-same as yesterday (-:

```
TASK [internetarchive : mkdir /opt/iiab/internetarchive] ************************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [internetarchive : Wipe dir /opt/iiab/internetarchive/node_modules (can be ~126+ MB) so './runrole --reinstall internetarchive' gets you the latest code] ******************************************************************************
changed: [127.0.0.1]

TASK [internetarchive : Run 'yarn add https://github.com/internetarchive/dweb-mirror' to download/populate /opt/iiab/internetarchive/node_modules (CAN TAKE ~5 MINUTES)] ********************************************************************
fatal: [127.0.0.1]: FAILED! => {"changed": true, "cmd": "yarn config set child-concurrency 1 && yarn add https://github.com/internetarchive/dweb-mirror", "delta": "0:01:23.001582", "end": "2023-09-12 06:08:04.177071", "msg": "non-zero return code", "rc": 127, "start": "2023-09-12 06:06:41.175489", "stderr": "warning https://github.com/internetarchive/dweb-mirror > multihashes > multibase@4.0.6: This module has been superseded by the multiformats module\nwarning https://github.com/internetarchive/dweb-mirror > @internetarchive/dweb-archive-dist > react > fbjs > core-js@1.2.7: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.\nwarning https://github.com/internetarchive/dweb-mirror > @futurepress/epubjs-reader > epubjs > @types/localforage@0.0.34: This is a stub types definition for localforage (https://github.com/localForage/localForage). localforage provides its own type definitions, so you don't need @types/localforage installed!\nerror /opt/iiab/internetarchive/node_modules/sharp: Command failed.\nExit code: 127\nCommand: (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)\nArguments: \nDirectory: /opt/iiab/internetarchive/node_modules/sharp\nOutput:\nsharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.14.4/libvips-8.14.4-linux-armv7.tar.br\nsharp: Integrity check passed for linux-armv7\nprebuild-install warn install No prebuilt binaries found (target=7 runtime=napi arch=arm libc= platform=linux)\n/bin/sh: 1: node-gyp: not found", "stderr_lines": ["warning https://github.com/internetarchive/dweb-mirror > multihashes > multibase@4.0.6: This module has been superseded by the multiformats module", "warning https://github.com/internetarchive/dweb-mirror > @internetarchive/dweb-archive-dist > react > fbjs > core-js@1.2.7: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.", "warning https://github.com/internetarchive/dweb-mirror > @futurepress/epubjs-reader > epubjs > @types/localforage@0.0.34: This is a stub types definition for localforage (https://github.com/localForage/localForage). localforage provides its own type definitions, so you don't need @types/localforage installed!", "error /opt/iiab/internetarchive/node_modules/sharp: Command failed.", "Exit code: 127", "Command: (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)", "Arguments: ", "Directory: /opt/iiab/internetarchive/node_modules/sharp", "Output:", "sharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.14.4/libvips-8.14.4-linux-armv7.tar.br", "sharp: Integrity check passed for linux-armv7", "prebuild-install warn install No prebuilt binaries found (target=7 runtime=napi arch=arm libc= platform=linux)", "/bin/sh: 1: node-gyp: not found"], "stdout": "yarn config v1.22.19\nsuccess Set \"child-concurrency\" to \"1\".\nDone in 0.12s.\nyarn add v1.22.19\ninfo No lockfile found.\n[1/4] Resolving packages...\n[2/4] Fetching packages...\n[3/4] Linking dependencies...\n[4/4] Building fresh packages...\ninfo Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.", "stdout_lines": ["yarn config v1.22.19", "success Set \"child-concurrency\" to \"1\".", "Done in 0.12s.", "yarn add v1.22.19", "info No lockfile found.", "[1/4] Resolving packages...", "[2/4] Fetching packages...", "[3/4] Linking dependencies...", "[4/4] Building fresh packages...", "info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command."]}
```

**Comment by mitra42** [comment]
I'm not sure - I was able to repeat the problem on my mac and fixed it, but that was just running on my mac. 
I also just did `yarn add https://github.com/internetarchive/dweb-mirror` successfully on my Mac.

I don't have another environment while I'm traveling to test it on.  

Were you trying on a fresh install.

**Comment by holta** [comment]
To simulate a fresh install, so far this works on "full 64-bit" but not on "hybrid 32-bit":

```
mv /opt/iiab/internetarchive /opt/iiab/internetarchive.old
cd /opt/iiab/iiab
./runrole internetarchive    # OR: ./runrole --reinstall internetarchive
```

(I just tried both again, "full 64-bit" Raspberry Pi OS works, whereas the same error recurs on their new-since-March "32-bit that actually boots a 64-bit kernel...")

Finally: If "32-bit" Raspberry Pi OS proves too complicated in the end, we definitely can consider dropping support for it, no worries!

**Comment by holta** [comment]
Fresh install data point, on a "true 32-bit kernel" Raspberry Pi OS 12 Bookworm — shows basically the same failure:

```
TASK [internetarchive : Run 'yarn add https://github.com/internetarchive/dweb-mirror' to download/populate /opt/iiab/internetarchive/node_modules (CAN TAKE ~5 MINUTES)] ***
fatal: [127.0.0.1]: FAILED! => {"changed": true, "cmd": "yarn config set child-concurrency 1 && yarn add https://github.com/internetarchive/dweb-mirror", "delta": "0:01:03.116691", "end": "2023-09-12 07:47:22.486828", "msg": "non-zero return code", "rc": 127, "start": "2023-09-12 07:46:19.370137", "stderr": "warning https://github.com/internetarchive/dweb-mirror > multihashes > multibase@4.0.6: This module has been superseded by the multiformats module\nwarning https://github.com/internetarchive/dweb-mirror > @internetarchive/dweb-archive-dist > react > fbjs > core-js@1.2.7: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.\nwarning https://github.com/internetarchive/dweb-mirror > @futurepress/epubjs-reader > epubjs > @types/localforage@0.0.34: This is a stub types definition for localforage (https://github.com/localForage/localForage). localforage provides its own type definitions, so you don't need @types/localforage installed!\nerror /opt/iiab/internetarchive/node_modules/sharp: Command failed.\nExit code: 127\nCommand: (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)\nArguments: \nDirectory: /opt/iiab/internetarchive/node_modules/sharp\nOutput:\nsharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.14.4/libvips-8.14.4-linux-armv7.tar.br\nsharp: Integrity check passed for linux-armv7\nprebuild-install warn install No prebuilt binaries found (target=7 runtime=napi arch=arm libc= platform=linux)\n/bin/sh: 1: node-gyp: not found", "stderr_lines": ["warning https://github.com/internetarchive/dweb-mirror > multihashes > multibase@4.0.6: This module has been superseded by the multiformats module", "warning https://github.com/internetarchive/dweb-mirror > @internetarchive/dweb-archive-dist > react > fbjs > core-js@1.2.7: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.", "warning https://github.com/internetarchive/dweb-mirror > @futurepress/epubjs-reader > epubjs > @types/localforage@0.0.34: This is a stub types definition for localforage (https://github.com/localForage/localForage). localforage provides its own type definitions, so you don't need @types/localforage installed!", "error /opt/iiab/internetarchive/node_modules/sharp: Command failed.", "Exit code: 127", "Command: (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)", "Arguments: ", "Directory: /opt/iiab/internetarchive/node_modules/sharp", "Output:", "sharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.14.4/libvips-8.14.4-linux-armv7.tar.br", "sharp: Integrity check passed for linux-armv7", "prebuild-install warn install No prebuilt binaries found (target=7 runtime=napi arch=arm libc= platform=linux)", "/bin/sh: 1: node-gyp: not found"], "stdout": "yarn config v1.22.19\nsuccess Set \"child-concurrency\" to \"1\".\nDone in 0.08s.\nyarn add v1.22.19\ninfo No lockfile found.\n[1/4] Resolving packages...\n[2/4] Fetching packages...\n[3/4] Linking dependencies...\n[4/4] Building fresh packages...\ninfo Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.", "stdout_lines": ["yarn config v1.22.19", "success Set \"child-concurrency\" to \"1\".", "Done in 0.08s.", "yarn add v1.22.19", "info No lockfile found.", "[1/4] Resolving packages...", "[2/4] Fetching packages...", "[3/4] Linking dependencies...", "[4/4] Building fresh packages...", "info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command."]}
```

Machine/OS system details:

```
root@raspberrypi:~# iiab-summary
install_date = 2023-09-11T01:47:34Z   Current TZ: EDT

iiab: 2804bf186, 51 PR's / 229 commits since tag 8.1-preview-2
    "Merge pull request #3643 from holta/ansible-core-2.15.4"
    https://github.com/iiab/iiab   branch: master

/etc/iiab/pr-list-pulled:
2023-09-10 21:45:51 EDT #3632 1 19861267 https://github.com/holta/iiab cryptography-experiment "scripts/ansible: Evaluate upstream piwheels / cryptography fix on 32-bit "RasPiOS 12" Lite"

iiab-admin-console: f085beb, 19 PR's / 62 commits since tag v0.6.2
    "Merge pull request #545 from tim-moody/0.6.6-job-outp"
    https://github.com/iiab/iiab-admin-console   branch: master

Raspberry Pi reference 2023-05-03
Generated using pi-gen, https://github.com/RPi-Distro/pi-gen, 47eee1f0ddcf8811559d51eea1c1bb48335e3e88, stage2
/etc/debian-version: 12.1
display-manager? inactive   Arch1: armhf   Arch2:
box 6.1.21-v7l+ #1642 SMP Mon Apr  3 17:22:30 BST 2023 armv7l
Cortex-A72   CPU(s): 4   Mem: 1871 Swap: 1023
Raspberry Pi 4 Model B Rev 1.1
Temperature(s): 53069

Filesystem      Size  Used Avail Use% Mounted on
/dev/root        59G  7.0G   51G  13% /      ZIMs: 1 OER2Go: 1 Apps2B: 1

lo eth0 wlan0 ap0 br0
openvpn_enabled: False
openvpn_handle: 182-raspios-armhf-armv7l-SMALL
192.168.0.182 10.10.10.10
```

**Comment by mitra42** [comment]
Yes - I'm guessing its something to do with pre-builds of shape not being available or something like that. 

Who does this impact, i.e. is it anyone with certain physical boxes, or is it just because someone has CHOSEN to use a 32 bit OS but could actually use a 64 bit one if they really wanted the internet archive tool?

**Comment by holta** [comment]
I agree 💯 %

Let's steer people away from 32-bit for now — and I'll check back in a month or two to see if upstream dependencies might work better then — and hopefully Raspberry Pi Foundation too, will start to formally steer people away from 32-bit for "full/real 64-bit" soon too!

Many thanks @mitra42 for having confirmed it's all working within your Mac / VM ✅

**Comment by holta** [comment]
@mitra42 just a clarification that https://piwheels.org is building Python 3.11 wheels for "32-bit" Raspberry Pi OS 12 / Bookworm as we speak (I didn't realize they weren't quite done yet).  So hypothetically this might also help in coming weeks?

_"We are working through the backlog now - it'll be another week maybe"_

- https://github.com/piwheels/packages/issues/391

**Comment by mitra42** [comment]
So I'm still unclear who this impacts - i.e. if its enforced by hardware they already own (e.g. if can only run 32 bit OS on certain hardware), or its a choice which OS (32 vs 64 bit)

**Comment by holta** [comment]
> enforced by hardware they already own (e.g. if can only run 32 bit OS on certain hardware)

That's definitely part of the dilemma, in some cases.

Another Part:

The Raspberry Pi Foundation / company heavily promotes right on top of their OS download page(s) a so-called "32-bit" OS — but then (suddenly and silently, since March!) this OS began inducing people to run 64-bit userspace / kernel — whenever certain more recent ARM CPU's are detected at boot time: raspberrypi/firmware#1795

In short:

A Raspberry Pi OS transition nudging everyone slowly-but-surely-year-by-year towards "full 64-bit" is clearly underway — with occasional rough spots along the road — but we'll know a lot more in coming weeks when the "Bookworm" version of Raspberry Pi OS becomes official. 🙃
